### PR TITLE
Enhance threat levels response time

### PR DESCRIPTION
**Pull request by tbartynski**

### CODE REVIEW

**Comment by mpawlik** [comment]
Do we need an assignee here?

**Comment by mkasztelnik** [comment]
For sure we need.

**Comment by tbartynski** [comment]
See comment in #151 
Closes #151

File: app/controllers/api/v1/threat_levels_controller.rb [hunk 1]
```diff
@@ -17,44 +17,83 @@ def set_limit
...
 
-      def threat_levels_for_profile(profile)
-        {
-          profile_id: profile.id,
-          threat_assessments: threat_assessments_for_profile(profile),
-          threat_level_assessment_runs: ThreatLevel::AssessmentRunsDataBuilderService.get(profile)
-        }
-      end
+      def fill_threat_assessments_data(results_data)
+        @threat_assessments.uniq!
```
Comment (nowakowski) [inline app/controllers/api/v1/threat_levels_controller.rb]: Kind of awkward. How about `@threat_assessments << ta unless @threat_assessments.include? ta` further on down (in `build_results_data:55`?

File: app/controllers/api/v1/threat_levels_controller.rb [hunk 2]
```diff
@@ -17,44 +17,83 @@ def set_limit
...
+        @threat_assessments.each do |ta|
+          profile = ta.profiles.first
```
Comment (nowakowski) [inline app/controllers/api/v1/threat_levels_controller.rb]: Wait, why `first`? Shouldn't this be an iteration over all `ta.profiles`?

File: app/controllers/api/v1/threat_levels_controller.rb [hunk 3]
```diff
@@ -8,7 +8,7 @@ class ThreatLevelsController < Api::ApiController
 
       def index
         set_limit
-        render json: threat_levels_data
+        render json: {threat_levels: threat_levels_for_profiles}
```
Comment (nowakowski) [inline app/controllers/api/v1/threat_levels_controller.rb]: I assume this syntax has been agreed upon with @mpawlik ?

File: app/services/threat_level/threat_assessment_runs_for_profile_service.rb [hunk 4]
```diff
@@ -1,8 +1,9 @@
 module ThreatLevel
   class ThreatAssessmentRunsForProfileService
-    def self.get(profile)
-      ThreatAssessmentRun.joins(threat_assessments: :profiles).
-          where("profiles.id = #{profile.id}")
+    def self.get(profile_id)
+      ThreatAssessmentRun.includes(threat_assessments: :results).
+        joins(threat_assessments: :profiles).
+          where("profiles.id = #{profile_id}").order(created_at: :desc).limit(5)
```
Comment (nowakowski) [inline app/services/threat_level/threat_assessment_runs_for_profile_service.rb]: We need a mention of this five-item limit in API docs.

File: app/controllers/api/v1/threat_levels_controller.rb (hunk 3 above)
Comment (tbartynski) [inline app/controllers/api/v1/threat_levels_controller.rb]: Response format has been agreed with Daniel and has not been changed in PR.

File: app/controllers/api/v1/threat_levels_controller.rb (hunk 1 above)
Comment (tbartynski) [inline app/controllers/api/v1/threat_levels_controller.rb]: I guess single invocation of uniq! is faster than checking the condition every time I want to add element. I can change it if you want me to.

File: app/services/threat_level/threat_assessment_runs_for_profile_service.rb (hunk 4 above)
Comment (tbartynski) [inline app/services/threat_level/threat_assessment_runs_for_profile_service.rb]: Right. Did that in commit https://github.com/ismop/dap/pull/156/commits/783f6e58d4172522dec9595aae54bb8ed0e4501c

File: app/controllers/api/v1/threat_levels_controller.rb (hunk 2 above)
Comment (tbartynski) [inline app/controllers/api/v1/threat_levels_controller.rb]: Good question. We've this conversation when I started working on _porównywaczka_ years (?) ago. We agreed to have such an assumption that assessment is associated with one profile despite model enables N-M associations. _Porównywaczka_  always creates an assessment associated with one profile. All assessments in prod DB have at most one profile.
Although you are right there is a potential bug in the code and I'll change it if you insist.

**Pull request merged**